application/states.go

```diff
@@ -0,0 +1,105 @@
+package application
+
+import (
+	"fmt"
+
+	"github.com/ONSdigital/dp-dataset-api/models"
+)
+
+type Created struct{}
+
+func (g Created) Enter(combinedVersionUpdate *models.Version, l *StateMachine) error {
+	fmt.Println("Creating")
+	return nil
+}
+func (g Created) String() string {
+	return "created"
+}
```

Comment:
Reflecting on this, I wonder if we used a struct with a func field then you could avoid multiple types which might be easier to use and more generic as the functions could have different receivers making it easier to instantiate.

For example, if the Struct type was changed from an interface to the following:

```go
type State struct {
	Name                string
	EnterFunc func(combinedVersionUpdate *models.Version, l *StateMachine) error
}

func (s State) String() string {
	return s.Name
}
```
Then each state could be defined like:

```suggestion
State{
	Name: "created"
	EnterFunc: application.CreateVersion
}
```

Reply:
Would these states still be defined in states.go or in service.go when the app starts up?

Reply:
Good question, I think the actual functions called should live in the application package, but the state configuration could live in either the application Setup or the service package and then just reference the enter functions in the application package. As the transitions are already configured in the service package currently, it having them there would be the most consistent. Logically however, as there is an application Setup() function and the transitions and states are all part of the application use cases, they would be neatly abstracted there.

Testing could be a good decider here by thinking about what we would and would not need to mock. If we think of the various testing scopes, when testing the state machine we would be testing it generically as we want to be able to change the states and transitions without changing the actual state machine so we would want a high level of test coverage and to pass in the transitions and states in the unit tests. When testing the application use cases, we are testing the actual app specific logic, therefore we would only want to mock repositories, event producers and other things in the interface adapters layer. The application rules around state change form part of the application use cases so having the transitions and states here makes sense and will be testable so long as we inject the interface adapters so they can be mocked.